File: checkout/java/src/test/java/org/pivotal/katas/CheckoutTest.java
package org.pivotal.katas;

import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;

public class CheckoutTest {
  private static final Map<String, Checkout.Pricing> RULES = new HashMap<String, Checkout.Pricing>() {{
    put("A", new Checkout.Pricing(50, new Checkout.Discount(3, 130)));
    put("B", new Checkout.Pricing(30, new Checkout.Discount(2, 45)));
    put("C", new Checkout.Pricing(20));
    put("D", new Checkout.Pricing(15));
  }};

  private int priceOf(String goods) {
    Checkout checkout = new Checkout(RULES);
    for(String item : goods.split("")) {
      checkout.scan(item);
    }
    return checkout.total();
  }

  @Test
  public void test_calculatesTotalsCorrectly() throws Exception {
    assertEquals(0, priceOf(""));
    assertEquals(50, priceOf("A"));
    assertEquals(80, priceOf("AB"));
    assertEquals(115, priceOf("CDBA"));

    assertEquals(100, priceOf("AA"));
    assertEquals(130, priceOf("AAA"));
    assertEquals(180, priceOf("AAAA"));
    assertEquals(230, priceOf("AAAAA"));
    assertEquals(260, priceOf("AAAAAA"));

    assertEquals(160, priceOf("AAAB"));
    assertEquals(175, priceOf("AAABB"));
    assertEquals(190, priceOf("AAABBD"));
    assertEquals(190, priceOf("DABABA"));
  }

  @Test
  public void test_calculatesTotalsIncrementally() throws Exception {
    Checkout checkout = new Checkout(RULES);
    assertEquals(0, checkout.total());
    checkout.scan("A"); assertEquals(50, checkout.total());
    checkout.scan("B"); assertEquals(80, checkout.total());
    checkout.scan("A"); assertEquals(130, checkout.total());
    checkout.scan("A"); assertEquals(160, checkout.total());
    checkout.scan("B"); assertEquals(175, checkout.total());
  }
}

File: checkout/ruby/test/checkout_test.rb
require 'minitest/autorun'
require_relative '../lib/checkout'

describe Checkout do
  RULES = {
      'A' => [50, [3, 130]],
      'B' => [30, [2, 45]],
      'C' => [20],
      'D' => [15]
  }

  def price_of(goods)
    checkout = Checkout.new(RULES)
    goods.split(//).each { |item| checkout.scan(item) }
    checkout.total
  end

  it 'calculates totals correctly' do
    price_of('').must_equal(0)
    price_of('A').must_equal(50)
    price_of('AB').must_equal(80)
    price_of('CDBA').must_equal(115)

    price_of('AA').must_equal(100)
    price_of('AAA').must_equal(130)
    price_of('AAAA').must_equal(180)
    price_of('AAAAA').must_equal(230)
    price_of('AAAAAA').must_equal(260)

    price_of('AAAB').must_equal(160)
    price_of('AAABB').must_equal(175)
    price_of('AAABBD').must_equal(190)
    price_of('DABABA').must_equal(190)
  end

  it 'calculates totals incrementally' do
    co = Checkout.new(RULES)
    co.total.must_equal(0)
    co.scan('A'); co.total.must_equal(50)
    co.scan('B'); co.total.must_equal(80)
    co.scan('A'); co.total.must_equal(130)
    co.scan('A'); co.total.must_equal(160)
    co.scan('B'); co.total.must_equal(175)
  end
end